bcda/client/bluebutton.go

```diff
@@ -54,7 +57,7 @@ func init() {
 	}
 }
 
-func NewBlueButtonClient() (*BlueButtonClient, error) {
+func NewBlueButtonClient(headerParams map[string]string) (*BlueButtonClient, error) {
```

Comment:
It would be preferable for the headers to be associated with individual requests rather than the client. The client could be used for many requests with different header values. I don't think we want it tied to one job, for instance.

Reply:
It is already tied to one job.  The job makes its own client and that client exists until the job is processed and then it gets garbage collected.  We didn't implement a singleton pattern for client so they are not shared between processes and we will often have several floating around in memory.  If the design changes dramatically we can adjust, but given the current state I don't think it makes any sense to push this down to the individual request level.  If an individual request needs to send different headers it can set them on the client as needed, or simply create a new client.  Doing it this way lets the requests be as generic as possible and doesn't force future devs to handle a bunch of low level things on every request, because they are just defined on the client.  I actually think the client should implement a singleton default headers so we aren't having to make them from scratch each time and we can just append the ones that change for each request.

Reply:
On review header is not an object the way URL params is so a singleton isn't really going to work.  However, I still maintain that this is largely a client concern and not a request concern.

Reply:
After reading through this code, I see both points. From a clean code perspective, the headers probably should be at the request level like @embh suggested. But to @knollfear point the way our code is currently implemented we create a client per job, clients aren't shared so the need to have it at the request level is nice but doesn't really give us anything since they all have separate clients anyway. If we were to change our implementation and make each request have a shared client object then the headers should absolutely be at the request level. And using knoll's idea of a default header and having the clients overwrite/change for their specific calls is a good one, but that should be combined with @embh idea at the request level, if we decide to change our design. Like getDefaultHeaders or something and each request could just overwrite with their own stuff...I think :)